### PR TITLE
[Java][Spring] fix one of

### PR DESCRIPTION
**Pull request by Orachigami**
Fixes: https://github.com/OpenAPITools/openapi-generator/issues/2906
Fixes: https://github.com/OpenAPITools/openapi-generator/issues/5381
Fixes: https://github.com/OpenAPITools/openapi-generator/issues/8647
Fixes: https://github.com/OpenAPITools/openapi-generator/issues/8860
Fixes: https://github.com/OpenAPITools/openapi-generator/issues/9601
Fixes: https://github.com/OpenAPITools/openapi-generator/issues/9981

Added tests `oneOfModelsGeneration` for Java Spring and Java Client which iterate through several specs with OneOf usage cases and check that OneOf models are generated and subtypes implement generated models. Java Client test also creates a cartesian product of specs and client libraries for regress tests.

### How inline OneOf models generation is fixed

Added `x-deduction` and `x-deduction-model-names` for `vendorExtensions`. These extensions are used to generated Jackson annotation when the discriminator cannot be used. The discriminator uses the specific field value for deserialization decisions. However, according to the https://swagger.io/specification#discriminator-object first example, `oneOf` can be used without the discriminator object. This case can be handled with Jackson recently added `JsonTypeInfo.Id.DEDUCTION` option. `DEDUCTION` option forces Jackson to make deserialization decisions according to the available fields in the payload and Models which implement `oneOf` interface/class.

### ToDo

* Polymorphic primitives deserialization

### Example output screenshots

<details>
  <summary>Output examples</summary>

<table>
  <tr>
    <th>Specification usage</th>
    <th>OneOf code</th>
    <th>Model code</th>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
   <tr>
      <td></td>
      <td><br></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>

</details>

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @nmuesch (2021/01)

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java [hunk 1]
```diff
@@ -1027,11 +1027,13 @@ public String toApiVarName(String name) {
...
-            Map<String, String> oneImport = new HashMap<>();
-            oneImport.put("import", importMapping.get(i));
-            if (!imports.contains(oneImport)) {
-                imports.add(oneImport);
+        if (additionalProperties.containsKey(JACKSON)) {
+            for (String i : Arrays.asList("JsonSubTypes", "JsonTypeInfo")) {
+                Map<String, String> oneImport = new HashMap<>();
+                oneImport.put("import", importMapping.get(i));
+                if (!imports.contains(oneImport)) {
+                    imports.add(oneImport);
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java:1035]: this sounds like `imports` should be a set

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java [hunk 2]
```diff
@@ -862,6 +864,19 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
         return objs;
     }
 
+    @Override
+    public void addImportsToOneOfInterface(List<Map<String, String>> imports) {
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:868]: is this method the same as the one in JavaClientCodegen?

Comment (Orachigami) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:868]: @agilob, implementation is the same for the present moment. I am not yet sure that I can extract the implementation to the `DefaultCodegen`.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java (hunk 1 above)
Comment (Orachigami) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java:1035]: @agilob, Set inside Map is not reliable. Any Map modifications will cause Set not to work properly.
```java
String key = "Wow";
String value = "hello";
HashSet<Map<String, String>> set = new HashSet<>() {{
    add(new HashMap<>());
}};
((HashMap<String,String>)set.toArray()[0]).put(key, value);
boolean result = set.add(new HashMap<>() {{
    put(key, value);
}});
System.err.println(result);
```

**Comment by Orachigami** [comment]
Moved to https://github.com/OpenAPITools/openapi-generator/pull/10463 as @wing suggested

**Pull request closed without merging**